File: packages/py/analytics/src/bolsa_analytics/cognitive/position_policy_decision.py
"""PositionPolicyDecision — autorización de policy post-entrada (V1.44).

≠ ExitPlan ≠ ExitPermission ≠ ExecutionPlan ≠ auto-exit.
No ejecuta. No muta PositionState. No llama al Router.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any, Literal

from bolsa_analytics.cognitive.exit_plan import ExitPlan, ExitReason
from bolsa_analytics.cognitive.exit_policy import suggestion_from_exit_policy
from bolsa_analytics.cognitive.operating_policy import OperatingPolicy
from bolsa_analytics.cognitive.position_event import (
    PositionEvent,
    build_position_event,
    is_immediate_risk_reason,
    is_target_event_kind,
    position_event_kind_from_reason,
)
from bolsa_analytics.cognitive.position_state import PositionState, clamp_stop_not_worsen

PositionPolicyVerdict = Literal["HOLD", "PROTECT", "TRAIL", "REDUCE", "EXIT"]
PositionPolicyRiskImpact = Literal["none", "reduce", "protect", "exit"]
PositionPolicyAuthorization = Literal["human_confirm", "policy"]
PositionPolicyDeferReason = Literal["queue_next_session", "data_stale"]

POSITION_POLICY_DECISION_KEY = "positionPolicyDecision"


def _now_iso(at: str | None = None) -> str:
    if isinstance(at, str) and at.strip():
        return at
    return datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%SZ")


@dataclass(frozen=True, slots=True)
class PositionPolicyDecision:
    verdict: PositionPolicyVerdict
    reason_code: ExitReason | None
    event: PositionEvent | None
    quantity: float | None
    new_stop: float | None
    target: float | None
    risk_impact: PositionPolicyRiskImpact
    policy_id: str
    as_of: str
    authorization: PositionPolicyAuthorization
    defer_reason: PositionPolicyDeferReason | None

    def to_dict(self) -> dict[str, Any]:
        return {
            "verdict": self.verdict,
            "reasonCode": self.reason_code,
            "event": (
                {
                    "kind": self.event.kind,
                    "reasonCode": self.event.reason_code,
                    "at": self.event.at,
                }
                if self.event
                else None
            ),
            "quantity": self.quantity,
            "newStop": self.new_stop,
            "target": self.target,
            "riskImpact": self.risk_impact,
            "policyId": self.policy_id,
            "asOf": self.as_of,
            "authorization": self.authorization,
            "deferReason": self.defer_reason,
        }


def _hold(
    policy: OperatingPolicy,
    as_of: str,
    reason_code: ExitReason | None,
    event: PositionEvent | None,
    defer_reason: PositionPolicyDeferReason | None,
) -> PositionPolicyDecision:
    return PositionPolicyDecision(
        verdict="HOLD",
        reason_code=reason_code,
        event=event,
        quantity=None,
        new_stop=None,
        target=None,
        risk_impact="none",
        policy_id=policy.template_id,
        as_of=as_of,
        authorization="policy",
        defer_reason=defer_reason,
    )


def _echo_target(position: PositionState, reason: str | None) -> float | None:
    if reason == "TARGET_1":
        return position.target1
    if reason == "TARGET_2":
        return position.target2
    return None


def decide_position_policy(
    position: PositionState | None,
    exit_plan: ExitPlan | None,
    operating_policy: OperatingPolicy,
    *,
    session: str | None = None,
    stale: bool | None = None,
    stop_touched: bool | None = None,
    as_of: str | None = None,
) -> PositionPolicyDecision:
    stamp = _now_iso(as_of or (exit_plan.updated_at if exit_plan else None))
    reason = exit_plan.primary_reason if exit_plan else None
    event = build_position_event(reason, stamp)
    kind = position_event_kind_from_reason(reason)

    if position is None or exit_plan is None or not reason or kind is None:
        return _hold(operating_policy, stamp, reason, event, None)

    immediate = is_immediate_risk_reason(reason) or stop_touched is True

    if stale is True and not immediate:
        return _hold(operating_policy, stamp, reason, event, "data_stale")

    if session == "closed" and is_target_event_kind(kind) and not immediate:
        return _hold(operating_policy, stamp, reason, event, "queue_next_session")

    action, qty, stop = suggestion_from_exit_policy(
        reason,
        position.remaining_quantity,
        operating_policy.exit,
        exit_plan.suggested_stop,
    )

    if action == "hold":
        return _hold(operating_policy, stamp, reason, event, None)

    new_stop = stop
    if new_stop is not None and new_stop > 0:
        new_stop = clamp_stop_not_worsen(
            position.direction, position.current_stop, new_stop
        )

    if action == "protect":
        trail = reason == "TRAIL"
        return PositionPolicyDecision(
            verdict="TRAIL" if trail else "PROTECT",
            reason_code=reason,
            event=event,
            quantity=None,
            new_stop=new_stop,
            target=_echo_target(position, reason),
            risk_impact="protect",
            policy_id=operating_policy.template_id,
            as_of=stamp,
            authorization="policy",
            defer_reason=None,
        )

    if action == "reduce":
        return PositionPolicyDecision(
            verdict="REDUCE",
            reason_code=reason,
            event=event,
            quantity=qty,
            new_stop=None,
            target=_echo_target(position, reason),
            risk_impact="reduce",
            policy_id=operating_policy.template_id,
            as_of=stamp,
            authorization="policy",
            defer_reason=None,
        )

    return PositionPolicyDecision(
        verdict="EXIT",
        reason_code=reason,
        event=event,
        quantity=qty if qty is not None else position.remaining_quantity,
        new_stop=None,
        target=_echo_target(position, reason),
        risk_impact="exit",
        policy_id=operating_policy.template_id,
        as_of=stamp,
        authorization="policy",
        defer_reason=None,
    )

File: packages/shared/src/cognitive/position-policy-decision.ts
/**
 * PositionPolicyDecision — autorización de policy post-entrada (V1.44).
 * ≠ ExitPlan ≠ ExitPermission ≠ ExecutionPlan ≠ auto-exit.
 * decidePositionPolicy no ejecuta, no muta PositionState, no llama al Router.
 */

import type { ExitPlanV1, ExitReasonV1 } from "./exit-plan.js";
import { suggestionFromExitPolicy } from "./exit-policy.js";
import type { OperatingPolicyV1 } from "./operating-policy.js";
import {
  buildPositionEvent,
  isImmediateRiskReason,
  isTargetEventKind,
  positionEventKindFromReason,
  type PositionEventV1,
} from "./position-event.js";
import { clampStopNotWorsen } from "./position-state.js";
import type { PositionStateV1 } from "./position-state.js";

export type PositionPolicyVerdictV1 =
  | "HOLD"
  | "PROTECT"
  | "TRAIL"
  | "REDUCE"
  | "EXIT";

export type PositionPolicyRiskImpactV1 = "none" | "reduce" | "protect" | "exit";

export type PositionPolicyAuthorizationV1 = "human_confirm" | "policy";

export type PositionPolicyDeferReasonV1 = "queue_next_session" | "data_stale";

export type PositionPolicyMarketContextV1 = {
  session?: "open" | "closed" | null;
  stale?: boolean | null;
  stopTouched?: boolean | null;
  asOf?: string | null;
};

export type PositionPolicyDecisionV1 = {
  verdict: PositionPolicyVerdictV1;
  reasonCode: ExitReasonV1 | null;
  event: PositionEventV1 | null;
  quantity: number | null;
  newStop: number | null;
  target: number | null;
  riskImpact: PositionPolicyRiskImpactV1;
  policyId: OperatingPolicyV1["templateId"];
  asOf: string;
  authorization: PositionPolicyAuthorizationV1;
  deferReason: PositionPolicyDeferReasonV1 | null;
};

export const POSITION_POLICY_DECISION_KEY = "positionPolicyDecision";

function nowIso(at?: string | null): string {
  if (typeof at === "string" && at.trim()) return at;
  return new Date().toISOString();
}

function hold(
  policy: OperatingPolicyV1,
  asOf: string,
  reasonCode: ExitReasonV1 | null,
  event: PositionEventV1 | null,
  deferReason: PositionPolicyDeferReasonV1 | null,
): PositionPolicyDecisionV1 {
  return {
    verdict: "HOLD",
    reasonCode,
    event,
    quantity: null,
    newStop: null,
    target: null,
    riskImpact: "none",
    policyId: policy.templateId,
    asOf,
    authorization: "policy",
    deferReason,
  };
}

function echoTarget(
  position: PositionStateV1,
  reason: ExitReasonV1 | null,
): number | null {
  if (reason === "TARGET_1") return position.target1;
  if (reason === "TARGET_2") return position.target2;
  return null;
}

/**
 * Policy authorization from current position + detected ExitPlan + OperatingPolicy.
 * Re-derives qty/stop from OperatingPolicy.exit (not TradePlan-time snapshot).
 */
export function decidePositionPolicy(
  position: PositionStateV1 | null | undefined,
  exitPlan: ExitPlanV1 | null | undefined,
  operatingPolicy: OperatingPolicyV1,
  marketContext?: PositionPolicyMarketContextV1 | null,
): PositionPolicyDecisionV1 {
  const ctx = marketContext ?? {};
  const asOf = nowIso(ctx.asOf ?? exitPlan?.updatedAt);
  const reason = exitPlan?.primaryReason ?? null;
  const event = buildPositionEvent(reason, asOf);
  const kind = positionEventKindFromReason(reason);

  if (!position || !exitPlan || !reason || !kind) {
    return hold(operatingPolicy, asOf, reason, event, null);
  }

  const immediate = isImmediateRiskReason(reason) || ctx.stopTouched === true;

  if (ctx.stale === true && !immediate) {
    return hold(operatingPolicy, asOf, reason, event, "data_stale");
  }

  if (ctx.session === "closed" && isTargetEventKind(kind) && !immediate) {
    return hold(operatingPolicy, asOf, reason, event, "queue_next_session");
  }

  const suggestion = suggestionFromExitPolicy(
    reason,
    position.remainingQuantity,
    operatingPolicy.exit,
    exitPlan.suggestedStop,
  );

  if (suggestion.suggestedAction === "hold") {
    return hold(operatingPolicy, asOf, reason, event, null);
  }

  let newStop: number | null = suggestion.suggestedStop;
  if (newStop != null && newStop > 0) {
    newStop = clampStopNotWorsen(
      position.direction,
      position.currentStop,
      newStop,
    );
  }

  if (suggestion.suggestedAction === "protect") {
    const trail = reason === "TRAIL";
    return {
      verdict: trail ? "TRAIL" : "PROTECT",
      reasonCode: reason,
      event,
      quantity: null,
      newStop,
      target: echoTarget(position, reason),
      riskImpact: "protect",
      policyId: operatingPolicy.templateId,
      asOf,
      authorization: "policy",
      deferReason: null,
    };
  }

  if (suggestion.suggestedAction === "reduce") {
    return {
      verdict: "REDUCE",
      reasonCode: reason,
      event,
      quantity: suggestion.suggestedQty,
      newStop: null,
      target: echoTarget(position, reason),
      riskImpact: "reduce",
      policyId: operatingPolicy.templateId,
      asOf,
      authorization: "policy",
      deferReason: null,
    };
  }

  return {
    verdict: "EXIT",
    reasonCode: reason,
    event,
    quantity: suggestion.suggestedQty ?? position.remainingQuantity,
    newStop: null,
    target: echoTarget(position, reason),
    riskImpact: "exit",
    policyId: operatingPolicy.templateId,
    asOf,
    authorization: "policy",
    deferReason: null,
  };
}
